spring-cloud-function-deployer/src/main/java/org/springframework/cloud/function/deployer/FunctionExtractingFunctionCatalog.java

```diff
@@ -165,13 +183,47 @@ public DeployedArtifact undeploy(String name) {
 			// TODO: Convert to 404
 			throw new IllegalStateException("No such app");
 		}
+		unregister(name);
 		this.deployer.undeploy(id);
 		String path = this.deployed.remove(id);
 		this.names.remove(name);
 		this.ids.remove(id);
 		return new DeployedArtifact(name, id, path);
 	}
 
+	private void register(String name) {
+		Set<String> names = getSupplierNames(name);
+		if (routes != null) {
+			logger.info("Registering routes: " + names);
+			routes.registerRoutes(getSupplierNames(name));
+		}
+		if (producer != null) {
+			// Need an ApplicationEvent that we can react to in the producer?
+			for (String supplier : names) {
+				producer.start(supplier);
+			}
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	private Set<String> getSupplierNames(String name) {
+		String id = this.names.get(name);
+		return (Set<String>) invoke(id, FunctionCatalog.class, "getSupplierNames");
+	}
+
+	private void unregister(String name) {
+		Set<String> names = getSupplierNames(name);
+		if (routes != null) {
+			logger.info("Unregistering routes: " + names);
+			routes.unregisterRoutes(names);
+		}
+		if (producer != null) {
+			for (String supplier : names) {
+				producer.stop(supplier);
```

Comment:
Not sure if it's relevant, but wondering if both `start/stop` calls should be wrapped into try/catch, otherwise failure of one call will negate all subsequent calls.